scipy/optimize/_hessian_update_strategy.py

```diff
@@ -0,0 +1,307 @@
+"""Implementation of Quasi-Newton update strategies."""
+
+from __future__ import division, print_function, absolute_import
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import get_blas_funcs
+
+
+class HessianUpdateStrategy(object):
+    """Quasi-Newton Hessian update strategy."""
```

Comment:
I have a feeling that this class should be fully documented and made publicly available. In theory a user might want to implement a new strategy just by subclassing this in his code. It should also contain a brief explanation that this is an abstract class and you should implement "update" method.

Do you agree?

Reply:
I agree with you.

Actually I think I should make a some change on my class structure to do what you are proposing. The major problem on the way it is currently implemented is that the class ``HessianUpdateStrategy`` contain some methods that are common to BFGS and SR1, but will not apply to all possible strategies. 

For instance, the methods ``initialize``, ``dot`` and ``get_matrix`` would have to be completely overwritten for a L-BFGS implementation.

So I propose to do the following: Create a "virtual" class ``HessianUpdateStrategy``:

```
class HessianUpdateStrategy(object):

    def initialize(self, n, approx_type):
        raise NotImplementedError()

    def update(self, delta_x, delta_grad):
        raise NotImplementedError()

    def dot(self, p):
        raise NotImplementedError()

    def get_matrix(self):
        raise NotImplementedError()
```

Than create a child class:
```
class FullHessianUpdateStrategy(HessianUpdateStrategy)
```
containing these methods that are common to BFGS and SR1 strategies.  And both classes BFGS and SR1 would inherit from this class.

This way I would make the "interface" ``HessianUpdateStrategy" public, giving freedom to the user to create its own custom update strategies with as much freedom as possible.